searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp

```diff
@@ -204,11 +204,84 @@ DenseTensorAttribute::extract_cells_ref(DocId docId) const
     }
     return _denseTensorStore.get_typed_cells(ref);
 }
+namespace {
+class Loader {
+public:
+    virtual ~Loader() = default;
+    virtual void load(uint32_t lid, vespalib::datastore::EntryRef ref) = 0;
+    virtual void wait_complete() = 0;
+};
+}
+
+
+class DenseTensorAttribute::ThreadedLoader : public Loader {
+public:
+    ThreadedLoader(DenseTensorAttribute & attr, vespalib::Executor & shared_executor)
+        : _attr(attr),
+          _shared_executor(shared_executor),
+          _complete_executor(1, 0x10000),
+          _pending(0)
+    {}
+    void load(uint32_t lid, vespalib::datastore::EntryRef ref) override {
+        {
+            std::unique_lock guard(_mutex);
+            while (_pending > MAX_PENDING) {
+                _cond.wait(guard);
+            }
+        }
+        while (_pending > 1000) { std::this_thread::yield(); }
+        ++_pending;
+        _shared_executor.execute(vespalib::makeLambdaTask([this, ref, lid]() {
+            auto prepared = _attr._index->prepare_add_document(lid, _attr._denseTensorStore.get_typed_cells(ref), _attr.getGenerationHandler().takeGuard());
+            _complete_executor.execute(vespalib::makeLambdaTask([this, lid, result=std::move(prepared)]() mutable {
```

Comment:
Note that this allows "out of order" completions. If we want "in order" completions (as during feeding) then the shared executor should populate a promise and the complete executor task should take a future as argument and be scheduled in the foreground (not by the shared executor task).

Reply:
Out of order is not a problem, as we do not have any ordering guarantees in vespa either. As ANN is in nature approximative that should be even less of an issue.
However I rewrote to use a queue and let the foreground thread do the completion too.
That avoids creating an extra thread and any issues with 2 threads modifying the attribute, even if they did seem to be disjunct.